tools/codegen/gen_functionalization_type.py

```diff
@@ -371,43 +379,68 @@ def emit_inplace_functionalization_body(
         {return_type} tmp_output;
         {{
           at::AutoDispatchSkipFunctionalize guard;
-          // The functionalization pass explicitly doesn't pass out= parameters to the redispatch
           {functional_call_str}
         }}
         {mutable_input_post_processing}
         {return_str(f)};
       }}
     }}"""
 
+# Generates the FunctionalizeAddBackViews kernels for:
+# - view_copy ops (which get turned back into view ops)
+def emit_view_copy_to_view_kernel(g: NativeFunctionsViewGroup) -> str:
+    view_sig = DispatcherSignature(g.view.func)
+    assert g.view_copy is not None
+    view_copy_sig = DispatcherSignature(g.view_copy.func)
 
-def emit_declaration_for_noncomposite_views(f: NativeFunction) -> str:
-    # For every view op, we need a corresponding "inverse view" function.
-    # This generates the declarations so we get a good compiler error when someone adds a new view.
-    view_inverse_sig = ViewInverseSignature(f)
-    return view_inverse_sig.decl()
+    view_api_name = g.view.func.name.unambiguous_name()
+    exprs = ', '.join([e.expr for e in translate(view_copy_sig.arguments(), view_sig.arguments())])
 
+    # The default generated composite kernel for {view}_copy() operators just clones
+    # the input tensor, and runs the underlying view on the clone.
+    return f"""
+// re-apply views "underneath" the functionalization pass, by converting view_copy calls to view calls.
+{view_copy_sig.defn(name=wrapper_name(g.view_copy.func), is_redispatching_fn=True)} {{
+  return at::_ops::{view_api_name}::call({exprs});
+}}
+"""
 
 # The below functions generate RegisterFunctionalization.cpp
 # These files provide the kernels that run the functionalization pass, which can be opted into
 # per backend (e.g. XLA or Vulkan), or as a composable transform (functionalize() in functorch).
 
-# TODO: later, I'll generate separate kernels for "Functionalize", and "AddBackViews"
-# That will probably be an enum, that this function will take in to know which ops get kernels.
 def needs_functionalization(
```

Comment:
as you are now unconditionally calling gen_ops_header, it doesn't seem to me like there is any good reason to have factored out the needs functionalization boolean test from the actual codegen. Just do the eligibility test in the codegen; it's not like you're saving any work splitting it up this way.